#!/usr/bin/env python
PACKAGE = "plane_segment"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

segment_criterion_enum = gen.enum([
                        gen.const("MAXIMUM_INLIER",       int_t, 0, ""),
                        gen.const("MINIMUM_CURVATURE",  int_t, 1, ""),
                        gen.const("BOTH",   int_t,  2,  "")
                       ], "Criterion to segment plane.")

gen.add("use_horizontal_line",  bool_t, 0,  "", True)
gen.add("use_verticle_line",  bool_t, 0,  "", True)
gen.add("y_skip",   int_t,  0,  "", 30, 1, 400)
gen.add("x_skip",   int_t,  0,  "", 30, 1, 400)
gen.add("line_point_min_distance",   double_t,   0,  "", 0.03, 0.001, 0.2)
gen.add("use_depth_noise_model",    bool_t, 0,  "", True)
gen.add("scan_rho_constant_error",  double_t,   0,  "", 0.01, 0.00001, 0.1)
gen.add("scan_rho_distance_error",  double_t,   0,  "", 0.004, -0.1, 0.1)
gen.add("scan_rho_quadratic_error", double_t,   0,  "", 0.004, 0.00001, 0.1)
gen.add("slide_window_size",  int_t,  0,  "", 11, 5, 23)
gen.add("line_min_inliers",  int_t,  0,  "", 30, 10,  200)
gen.add("line_fitting_threshold",   double_t,   0,  "", 1.2, 0.01, 10.0)
#
gen.add("normals_per_line", int_t,  0,  "", 1,  1,  11)
gen.add("normal_use_depth_dependent_smoothing",  bool_t, 0,  "", False)
gen.add("normal_max_depth_change_factor",   double_t,   0,  "", 0.02, 0.001, 0.2)
gen.add("normal_smoothing_size",    int_t,  0,  "", 30, 5, 50)
gen.add("normal_min_inliers_percentage",    double_t,  0,  "", 0.6, 0.2, 0.99)
gen.add("normal_maximum_curvature", double_t,   0,  "", 0.02, 0.001, 0.5)
#
gen.add("remove_duplicate_candidate",  bool_t, 0,  "", True)
gen.add("duplicate_candidate_normal_thresh", double_t,   0,  "", 0.08, 1e-4, 0.5)
gen.add("duplicate_candidate_distance_thresh", double_t,   0,  "", 0.02, 1e-4, 0.5)
#
gen.add("plane_segment_criterion",  int_t,  0,  "", 0, 0, 2, edit_method=segment_criterion_enum)
gen.add("k_curvature",  double_t,   0,  "", 50.0, 1e-8, 10000)
gen.add("k_inlier",  double_t,   0,  "", 1e-5, 1e-8, 10000)
gen.add("distance_threshold", double_t,   0,  "", 0.02, 0.001, 0.20)
gen.add("neighbor_threshold",  double_t,   0,  "", 0.03, 0.001, 0.1)
gen.add("min_inliers", int_t,  0,  "", 5000, 100, 40000)
gen.add("max_curvature", double_t,   0,  "", 0.02, 0.001, 0.5)
gen.add("optimize_coefficients",  bool_t, 0,  "", True)
gen.add("project_points",  bool_t, 0,  "", False)
gen.add("extract_boundary",  bool_t, 0,  "", True)
##
gen.add("is_extract_single_plane",  bool_t, 0,  "", False)
gen.add("single_plane_row",    int_t,  0,  "", 200, 0, 479)
gen.add("single_plane_col",    int_t,  0,  "", 200, 0, 639)


exit(gen.generate(PACKAGE, "plane_segment", "LineBasedSegment"))
